### PR TITLE
koord-scheduler: support debug scores

### PR DESCRIPTION
**Pull request by eahydra**
koord-scheduler: support debug scores

Users can enable debug scores via flags -s, --debug-scores or set on demand via the following commands:

curl -X PUT leaderSchedulerIP:port/debug/flags/s --data '100'

Signed-off-by: Joseph <joseph.t.lee@outlook.com>

### Ⅰ. Describe what this PR does

Users can enable debug scores via flags -s, --debug-scores or set on demand via the following commands:

```bash
curl -X PUT <leaderSchedulerIP>:<port>/debug/flags/s --data '100'
```

and the scheduler will logging the following content with markdown style:

```
| # | Pod | Node | Score | ImageLocality | InterPodAffinity | LoadAwareScheduling | NodeAffinity | NodeNUMAResource | NodeResourcesBalancedAllocation | NodeResourcesFit | PodTopologySpread | Reservation | TaintToleration |
| --- | --- | --- | ---:| ---:| ---:| ---:| ---:| ---:| ---:| ---:| ---:| ---:| ---:|
| 0 | default/curlimage-545745d8f8-rngp7 | cn-hangzhou.10.0.4.51 | 577 | 0 | 0 | 87 | 0 | 0 | 96 | 94 | 200 | 0 | 100 |
| 1 | default/curlimage-545745d8f8-rngp7 | cn-hangzhou.10.0.4.50 | 574 | 0 | 0 | 85 | 0 | 0 | 96 | 93 | 200 | 0 | 100 |
| 2 | default/curlimage-545745d8f8-rngp7 | cn-hangzhou.10.0.4.19 | 541 | 0 | 0 | 55 | 0 | 0 | 95 | 91 | 200 | 0 | 100 |
| 3 | default/curlimage-545745d8f8-rngp7 | cn-hangzhou.10.0.4.18 | 487 | 0 | 0 | 15 | 0 | 0 | 90 | 82 | 200 | 0 | 100 |
```

| # | Pod | Node | Score | ImageLocality | InterPodAffinity | LoadAwareScheduling | NodeAffinity | NodeNUMAResource | NodeResourcesBalancedAllocation | NodeResourcesFit | PodTopologySpread | Reservation | TaintToleration |
| --- | --- | --- | ---:| ---:| ---:| ---:| ---:| ---:| ---:| ---:| ---:| ---:| ---:|
| 0 | default/curlimage-545745d8f8-rngp7 | cn-hangzhou.10.0.4.51 | 577 | 0 | 0 | 87 | 0 | 0 | 96 | 94 | 200 | 0 | 100 |
| 1 | default/curlimage-545745d8f8-rngp7 | cn-hangzhou.10.0.4.50 | 574 | 0 | 0 | 85 | 0 | 0 | 96 | 93 | 200 | 0 | 100 |
| 2 | default/curlimage-545745d8f8-rngp7 | cn-hangzhou.10.0.4.19 | 541 | 0 | 0 | 55 | 0 | 0 | 95 | 91 | 200 | 0 | 100 |
| 3 | default/curlimage-545745d8f8-rngp7 | cn-hangzhou.10.0.4.18 | 487 | 0 | 0 | 15 | 0 | 0 | 90 | 82 | 200 | 0 | 100 |

### Ⅱ. Does this pull request fix one issue?

implements #462 

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/scheduler/frameworkext/framework_extender.go [hunk 1]
```diff
@@ -208,6 +208,14 @@ func (ext *frameworkExtenderImpl) RunFilterPluginsWithNominatedPods(ctx context.
 	return ext.Framework.RunFilterPluginsWithNominatedPods(ctx, cycleState, pod, nodeInfo)
 }
 
+func (ext *frameworkExtenderImpl) RunScorePlugins(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodes []*corev1.Node) (framework.PluginToNodeScores, *framework.Status) {
+	pluginToNodeScores, status := ext.handle.RunScorePlugins(ctx, state, pod, nodes)
+	if status.IsSuccess() && debugTopNScores > 0 {
+		debugScores(debugTopNScores, pod, pluginToNodeScores, nodes)
```
Comment (saintube) [inline pkg/scheduler/frameworkext/framework_extender.go:214]: /lgtm

**Review by saintube — APPROVED** [review]

File: pkg/util/routes/flags.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+// StringFlagSetterFunc is a func used for setting string type flag.
+type StringFlagSetterFunc func(string) (string, error)
+
+// StringFlagPutHandler wraps an http Handler to set string type flag.
+func StringFlagPutHandler(setter StringFlagSetterFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		switch {
+		case req.Method == "PUT":
+			body, err := ioutil.ReadAll(req.Body)
```
Comment (jasonliu747) [inline pkg/util/routes/flags.go:108]: use `io.ReadAll` instead

**Review by jasonliu747 — CHANGES_REQUESTED** [review]

File: pkg/util/routes/flags.go (hunk 2 above)
Comment (eahydra) [reply pkg/util/routes/flags.go]: fixed

**Review by jasonliu747 — APPROVED** [review]
/lgtm

**Comment by hormes** [comment]
Should this be put into a separate file, so that it may be possible to read this file through the http interface in the future?

**Comment by eahydra** [comment]
> Should this be put into a separate file, so that it may be possible to read this file through the http interface in the future?

Very good advice. In fact, I have also considered this problem, but I want to provide a simple method to meet the needs of the most basic debug scores. Later, we will look at the feedback and needs of the community before deciding how to persist and query the results.

**Pull request merged**